sdk/cosmos/azure-cosmos/azure/cosmos/container.py

```diff
@@ -58,86 +67,78 @@ class ContainerProxy(object):
+    def __init__(
+            self,
+            client_connection: CosmosClientConnection,
+            database_link: str,
+            id: str,
+            properties: Optional[Dict[str, Any]] = None
+        ) -> None:
         self.id = id
+        self.container_link = "{}/colls/{}".format(database_link, self.id)
+        self.client_connection = client_connection
         self._properties = properties
-        self.container_link = u"{}/colls/{}".format(database_link, self.id)
-        self._is_system_key = None
-        self._scripts = None  # type: Optional[ScriptsProxy]
+        self._is_system_key: Optional[bool] = None
+        self._scripts: Optional[ScriptsProxy] = None
 
-    def __repr__(self):
-        # type () -> str
+    def __repr__(self) -> str:
         return "<ContainerProxy [{}]>".format(self.container_link)[:1024]
 
-    def _get_properties(self):
-        # type: () -> Dict[str, Any]
+    def _get_properties(self) -> Dict[str, Any]:
         if self._properties is None:
             self._properties = self.read()
         return self._properties
 
     @property
-    def is_system_key(self):
-        # type: () -> bool
+    def is_system_key(self) -> bool:
         if self._is_system_key is None:
             properties = self._get_properties()
             self._is_system_key = (
                 properties["partitionKey"]["systemKey"] if "systemKey" in properties["partitionKey"] else False
             )
-        return cast('bool', self._is_system_key)
+        return self._is_system_key
 
     @property
-    def scripts(self):
-        # type: () -> ScriptsProxy
+    def scripts(self) -> ScriptsProxy:
         if self._scripts is None:
             self._scripts = ScriptsProxy(self.client_connection, self.container_link, self.is_system_key)
-        return cast('ScriptsProxy', self._scripts)
+        return self._scripts
 
-    def _get_document_link(self, item_or_link):
-        # type: (Union[Dict[str, Any], str]) -> str
+    def _get_document_link(self, item_or_link: Union[str, Mapping[str, Any]]) -> str:
         if isinstance(item_or_link, str):
-            return u"{}/docs/{}".format(self.container_link, item_or_link)
+            return "{}/docs/{}".format(self.container_link, item_or_link)
         return item_or_link["_self"]
 
-    def _get_conflict_link(self, conflict_or_link):
-        # type: (Union[Dict[str, Any], str]) -> str
+    def _get_conflict_link(self, conflict_or_link: Union[str, Mapping[str, Any]]) -> str:
         if isinstance(conflict_or_link, str):
-            return u"{}/conflicts/{}".format(self.container_link, conflict_or_link)
+            return "{}/conflicts/{}".format(self.container_link, conflict_or_link)
         return conflict_or_link["_self"]
 
-    def _set_partition_key(self, partition_key):
+    def _set_partition_key(
+        self,
+        partition_key: Union[str, int, float, bool, List[Union[str, int, float, bool]]]
```

Comment:
would `_Empty` & `_Undefined` be valid values here ?

Reply:
These are only meant to be returned, not used by the user - however NonePartitionKeyValue would be valid (we even check for it below)

Reply:
This one's a bit awkward - because somehow we ended up with a really weird unPythonic design here where a class  definition is being used a value. This can technically be typed as `Type[NonePartitionKeyValue]`, however this really doesn't play well with mypy at all.
Ideally I would change this to use the singleton-enum structure here:
https://peps.python.org/pep-0484/#support-for-singleton-types-in-unions

hmmm I think I will try to address this in a separate PR as it might need some discussion.

Reply:
Nevermind - I've just typed as it's currently used (`Type[NonePartitionKeyValue]`) and used `cast` to avoid the warnings for now.